### PR TITLE
cql3: detect and handle int overflow in aggregate functions

### LINKED ISSUE #5536: cql3/functions/aggregate_fcts.hh:107:14: runtime error: signed integer overflow: 1446437040 + 732870167 cannot be represented in type 'int'
Scylla version 9d454b7dc6c7bb23f7044e8b7bf918ea9bf429b4
Reloc package: http://downloads.scylladb.com.s3.amazonaws.com/relocatable/unstable/master/201912302059/scylla-debug-package.tar.gz

Seen in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-debug-random/36/artifact/logs-debug.2/1577760890727_paging_test.TestPagingWithIndexingAndAggregation.test_filter_pk_column/node1.log
```
cql3/functions/aggregate_fcts.hh:107:14: runtime error: signed integer overflow: 1446437040 + 732870167 cannot be represented in type 'int'
Aborting on shard 1.
Backtrace:
  /jenkins/workspace/scylla-master/dtest-debug-random/scylla/.ccm/scylla-repository/9d454b7dc6c7bb23f7044e8b7bf918ea9bf429b4/scylla//opt/scylladb/libreloc/libasan.so.5+0x000000000006ce7f
  0x000000000a121d48
  0x000000000a0fddfc
  0x0000000009fa513b
  0x0000000009fa5281
  0x0000000009ffbb48
  0x000000000a037e4d
  0x000000000a037ece
  0x00007ffb42656b1f
  /jenkins/workspace/scylla-master/dtest-debug-random/scylla/.ccm/scylla-repository/9d454b7dc6c7bb23f7044e8b7bf918ea9bf429b4/scylla//opt/scylladb/libreloc/libc.so.6+0x000000000003c624
  /jenkins/workspace/scylla-master/dtest-debug-random/scylla/.ccm/scylla-repository/9d454b7dc6c7bb23f7044e8b7bf918ea9bf429b4/scylla//opt/scylladb/libreloc/libc.so.6+0x00000000000258d8
  /jenkins/workspace/scylla-master/dtest-debug-random/scylla/.ccm/scylla-repository/9d454b7dc6c7bb23f7044e8b7bf918ea9bf429b4/scylla//opt/scylladb/libreloc/libubsan.so.1+0x000000000001c1e1
  /jenkins/workspace/scylla-master/dtest-debug-random/scylla/.ccm/scylla-repository/9d454b7dc6c7bb23f7044e8b7bf918ea9bf429b4/scylla//opt/scylladb/libreloc/libubsan.so.1+0x0000000000026d8b
  /jenkins/workspace/scylla-master/dtest-debug-random/scylla/.ccm/scylla-repository/9d454b7dc6c7bb23f7044e8b7bf918ea9bf429b4/scylla//opt/scylladb/libreloc/libubsan.so.1+0x0000000000007e98
  /jenkins/workspace/scylla-master/dtest-debug-random/scylla/.ccm/scylla-repository/9d454b7dc6c7bb23f7044e8b7bf918ea9bf429b4/scylla//opt/scylladb/libreloc/libubsan.so.1+0x000000000000a71e
  /jenkins/workspace/scylla-master/dtest-debug-random/scylla/.ccm/scylla-repository/9d454b7dc6c7bb23f7044e8b7bf918ea9bf429b4/scylla//opt/scylladb/libreloc/libubsan.so.1+0x000000000000bf92
  0x000000000331e23a
  0x00000000048b5fe1
  0x000000000490ac77
  0x00000000048f53cb
  0x00000000048f54f0
  0x0000000002fd5fb9
  0x0000000002fd5a7e
  0x000000000732b937
  0x000000000731e554
  0x00000000073161d1
  0x00000000072f8d85
  0x0000000007303431
  0x00000000072ff3b8
  0x00000000072f8f6c
  0x00000000072f922a
  0x00000000072f9a3e
  0x00000000073036ca
  0x0000000007303853
  0x00000000073039c8
  0x00000000072ff571
  0x000000000730940f
  0x00000000073365b8
  0x000000000732c0e4
  0x000000000732c1e1
  0x000000000732c356
  0x000000000731f33d
  0x00000000073434e4
  0x0000000009fe19f2
  0x0000000009fe5b5b
  0x0000000009fead08
  0x0000000009ffe5a5
  0x000000000a054595
  0x0000000002a95fac
  0x000000000ad083be
  /jenkins/workspace/scylla-master/dtest-debug-random/scylla/.ccm/scylla-repository/9d454b7dc6c7bb23f7044e8b7bf918ea9bf429b4/scylla//opt/scylladb/libreloc/libpthread.so.0+0x00000000000094e1
  /jenkins/workspace/scylla-master/dtest-debug-random/scylla/.ccm/scylla-repository/9d454b7dc6c7bb23f7044e8b7bf918ea9bf429b4/scylla//opt/scylladb/libreloc/libc.so.6+0x0000000000101692

cql3::functions::aggregate_fcts::impl_sum_function_for<int>::add_input(cql_serialization_format, std::vector<std::optional<seastar::basic_sstring<signed char, unsigned int, 31u, false> >, std::allocator<std::optional<seastar::basic_sstring<signed char, unsigned int, 31u, false> > > > const&) at /jenkins/workspace/scylla-master/build/scylla/cql3/functions/aggregate_fcts.hh:107 (discriminator 6)
cql3::selection::aggregate_function_selector::add_input(cql_serialization_format, cql3::selection::result_set_builder&) at /jenkins/workspace/scylla-master/build/scylla/cql3/selection/aggregate_function_selector.hh:65
cql3::selection::selection_with_processing::selectors_with_processing::add_input_row(cql_serialization_format, cql3::selection::result_set_builder&) at /jenkins/workspace/scylla-master/build/scylla/cql3/selection/selection.cc:214 (discriminator 2)
cql3::selection::result_set_builder::process_current_row(bool) at /jenkins/workspace/scylla-master/build/scylla/cql3/selection/selection.cc:363
cql3::selection::result_set_builder::new_row() at /jenkins/workspace/scylla-master/build/scylla/cql3/selection/selection.cc:372
cql3::selection::result_set_builder::visitor<cql3::selection::result_set_builder::nop_filter>::accept_new_row(query::result_row_view const&, query::result_row_view const&) at /jenkins/workspace/scylla-master/build/scylla/./cql3/selection/selection.hh:397
cql3::selection::result_set_builder::visitor<cql3::selection::result_set_builder::nop_filter>::accept_new_row(clustering_key_prefix const&, query::result_row_view const&, query::result_row_view const&) at /jenkins/workspace/scylla-master/build/scylla/./cql3/selection/selection.hh:388
service::pager::query_pager::query_result_visitor<cql3::selection::result_set_builder::visitor<cql3::selection::result_set_builder::nop_filter> >::accept_new_row(clustering_key_prefix const&, query::result_row_view const&, query::result_row_view const&) at /jenkins/workspace/scylla-master/build/scylla/service/pager/query_pagers.cc:292
void query::result_view::consume<service::pager::query_pager::query_result_visitor<cql3::selection::result_set_builder::visitor<cql3::selection::result_set_builder::nop_filter> >&>(query::partition_slice const&, service::pager::query_pager::query_result_visitor<cql3::selection::result_set_builder::visitor<cql3::selection::result_set_builder::nop_filter> >&) const at /jenkins/workspace/scylla-master/build/scylla/./query-result-reader.hh:190 (discriminator 1)
void service::pager::query_pager::handle_result<cql3::selection::result_set_builder::visitor<cql3::selection::result_set_builder::nop_filter> >(cql3::selection::result_set_builder::visitor<cql3::selection::result_set_builder::nop_filter>&&, seastar::foreign_ptr<seastar::lw_shared_ptr<query::result> > const&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >) at /jenkins/workspace/scylla-master/build/scylla/service/pager/query_pagers.cc:326
service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}::operator()(service::storage_proxy::coordinator_query_result) const::{lambda()#1}::operator()() const at /jenkins/workspace/scylla-master/build/scylla/service/pager/query_pagers.cc:200
seastar::future<> seastar::do_void_futurize_helper<void>::apply<service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}::operator()(service::storage_proxy::coordinator_query_result) const::{lambda()#1}>(service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}::operator()(service::storage_proxy::coordinator_query_result) const::{lambda()#1}&&) at /jenkins/workspace/scylla-master/build/scylla/seastar/include/seastar/core/future.hh:1507
seastar::future<> seastar::futurize<void>::apply<service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}::operator()(service::storage_proxy::coordinator_query_result) const::{lambda()#1}>(service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}::operator()(service::storage_proxy::coordinator_query_result) const::{lambda()#1}&&) at /jenkins/workspace/scylla-master/build/scylla/seastar/include/seastar/core/future.hh:1556
auto seastar::futurize_apply<service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}::operator()(service::storage_proxy::coordinator_query_result) const::{lambda()#1}>(service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}::operator()(service::storage_proxy::coordinator_query_result) const::{lambda()#1}&&) at /jenkins/workspace/scylla-master/build/scylla/seastar/include/seastar/core/future.hh:1645
auto cql3::selection::result_set_builder::with_thread_if_needed<service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}::operator()(service::storage_proxy::coordinator_query_result) const::{lambda()#1}>(service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}::operator()(service::storage_proxy::coordinator_query_result) const::{lambda()#1}&&) at /jenkins/workspace/scylla-master/build/scylla/./cql3/selection/selection.hh:266
service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}::operator()(service::storage_proxy::coordinator_query_result) const at /jenkins/workspace/scylla-master/build/scylla/service/pager/query_pagers.cc:202
seastar::apply_helper<service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, std::tuple<service::storage_proxy::coordinator_query_result&&>&&, std::integer_sequence<unsigned long, 0ul> >::apply(service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}, std::tuple<service::storage_proxy::coordinator_query_result&&>) at /jenkins/workspace/scylla-master/build/scylla/seastar/include/seastar/core/apply.hh:36
auto seastar::apply<service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, service::storage_proxy::coordinator_query_result&&>(service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, std::tuple<service::storage_proxy::coordinator_query_result&&>&&) at /jenkins/workspace/scylla-master/build/scylla/seastar/include/seastar/core/apply.hh:44
seastar::future<> seastar::futurize<seastar::future<> >::apply<service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, service::storage_proxy::coordinator_query_result&&>(service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, std::tuple<service::storage_proxy::coordinator_query_result&&>&&) at /jenkins/workspace/scylla-master/build/scylla/seastar/include/seastar/core/future.hh:1563
seastar::future<> seastar::future<service::storage_proxy::coordinator_query_result>::then<service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}, seastar::future<> >(service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&&)::{lambda(service::storage_proxy::coordinator_query_result&&)#1}::operator()(service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&&) at /jenkins/workspace/scylla-master/build/scylla/seastar/include/seastar/core/future.hh:1094
seastar::noncopyable_function<seastar::future<> (service::storage_proxy::coordinator_query_result&&)>::direct_vtable_for<seastar::future<> seastar::future<service::storage_proxy::coordinator_query_result>::then<service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}, seastar::future<> >(service::pager::query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&&)::{lambda(service::storage_proxy::coordinator_query_result&&)#1}>::call(seastar::noncopyable_function<seastar::future<> (service::storage_proxy::coordinator_query_result&&)> const*, service::storage_proxy::coordinator_query_result&&) at /jenkins/workspace/scylla-master/build/scylla/seastar/include/seastar/util/noncopyable_function.hh:99
seastar::noncopyable_function<seastar::future<> (service::storage_proxy::coordinator_query_result&&)>::operator()(service::storage_proxy::coordinator_query_result&&) const at /jenkins/workspace/scylla-master/build/scylla/seastar/include/seastar/util/noncopyable_function.hh:187
seastar::apply_helper<seastar::noncopyable_function<seastar::future<> (service::storage_proxy::coordinator_query_result&&)>, std::tuple<service::storage_proxy::coordinator_query_result>&&, std::integer_sequence<unsigned long, 0ul> >::apply(seastar::noncopyable_function<seastar::future<> (service::storage_proxy::coordinator_query_result&&)>&&, std::tuple<service::storage_proxy::coordinator_query_result>&&) at /jenkins/workspace/scylla-master/build/scylla/seastar/include/seastar/core/apply.hh:36
auto seastar::apply<seastar::noncopyable_function<seastar::future<> (service::storage_proxy::coordinator_query_result&&)>, service::storage_proxy::coordinator_query_result>(seastar::noncopyable_function<seastar::future<> (service::storage_proxy::coordinator_query_result&&)>&&, std::tuple<service::storage_proxy::coordinator_query_result>&&) at /jenkins/workspace/scylla-master/build/scylla/seastar/include/seastar/core/apply.hh:44
seastar::future<> seastar::futurize<seastar::future<> >::apply<seastar::noncopyable_function<seastar::future<> (service::storage_proxy::coordinator_query_result&&)>, service::storage_proxy::coordinator_query_result>(seastar::noncopyable_function<seastar::future<> (service::storage_proxy::coordinator_query_result&&)>&&, std::tuple<service::storage_proxy::coordinator_query_result>&&) at /jenkins/workspace/scylla-master/build/scylla/seastar/include/seastar/core/future.hh:1563
seastar::future<> seastar::future<service::storage_proxy::coordinator_query_result>::then_impl<seastar::noncopyable_function<seastar::future<> (service::storage_proxy::coordinator_query_result&&)>, seastar::future<> >(seastar::noncopyable_function<seastar::future<> (service::storage_proxy::coordinator_query_result&&)>&&)::{lambda()#1}::operator()() const::{lambda(seastar::future_state<service::storage_proxy::coordinator_query_result>&&)#1}::operator()(seastar::future_state<service::storage_proxy::coordinator_query_result>) at /jenkins/workspace/scylla-master/build/scylla/seastar/include/seastar/core/future.hh:1122
seastar::continuation<seastar::future<> seastar::future<service::storage_proxy::coordinator_query_result>::then_impl<seastar::noncopyable_function<seastar::future<> (service::storage_proxy::coordinator_query_result&&)>, seastar::future<> >(seastar::noncopyable_function<seastar::future<> (service::storage_proxy::coordinator_query_result&&)>&&)::{lambda()#1}::operator()() const::{lambda(seastar::future_state<service::storage_proxy::coordinator_query_result>&&)#1}, service::storage_proxy::coordinator_query_result>::run_and_dispose() at /jenkins/workspace/scylla-master/build/scylla/seastar/include/seastar/core/future.hh:450
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at /jenkins/workspace/scylla-master/build/scylla/seastar/src/core/reactor.cc:2106
seastar::reactor::run_some_tasks() at /jenkins/workspace/scylla-master/build/scylla/seastar/src/core/reactor.cc:2530 (discriminator 2)
seastar::reactor::run() at /jenkins/workspace/scylla-master/build/scylla/seastar/src/core/reactor.cc:2678
seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::{lambda()#3}::operator()() const at /jenkins/workspace/scylla-master/build/scylla/seastar/src/core/reactor.cc:3814 (discriminator 3)
std::_Function_handler<void (), seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::{lambda()#3}>::_M_invoke(std::_Any_data const&) at /usr/include/c++/9/bits/std_function.h:300
std::function<void ()>::operator()() const at /usr/include/c++/9/bits/std_function.h:690
seastar::posix_thread::start_routine(void*) at /jenkins/workspace/scylla-master/build/scylla/seastar/src/core/posix.cc:52
```

The statement that may hit integer overflow is:
https://github.com/scylladb/scylla/blob/9d454b7dc6c7bb23f7044e8b7bf918ea9bf429b4/cql3/functions/aggregate_fcts.hh#L107

Cassandra 3.11.3 doesn't return any error if sum() overflows so we should allow that and just print a warning to the log.

### PR DESCRIPTION
**Pull request by bhalevy**
Fix overflow handling in sum() and avg().

sum:
- aggregated into __int128
- detect overflow when computing result
  - and log a warning if found

avg:
- fix division function to divide the accumulator type _sum (__int128 for integers) by _count

- Add unit tests for both cases

Test:
- manual test against Cassandra 3.11.3 to make sure the results in the scylla unit test agree with it.
- unit(dev), cql_query_test(debug)

Fixes #5536

### CODE REVIEW

File: cql3/functions/functions.cc [hunk 1]
```diff
@@ -684,6 +684,20 @@ function_call::raw::to_string() const {
     return format("{}({})", _name, join(", ", _terms));
 }
 
+namespace aggregate_fcts {
+
+template <typename Type>
+void impl_sum_function_for<Type>::validate_sum() {
+    Type short_sum = static_cast<Type>(_sum);
+    if (static_cast<accumulator_type>(short_sum) != _sum) {
+        log.warn(format("SUM overflow: 0x{:08x}_{:08x}_{:08x}_{:08x} -> {}",
```
Comment (espindola) [inline cql3/functions/functions.cc]: My main question is: Do we need this? Does cassandra produce the same warning?

If in CQL signed integer arithmetic is defined to be modular, why not just use std::make_unsigned to select the type and skip the warning?

File: cql3/functions/functions.cc [hunk 2]
```diff
@@ -684,6 +684,20 @@ function_call::raw::to_string() const {
     return format("{}({})", _name, join(", ", _terms));
 }
 
+namespace aggregate_fcts {
+
+template <typename Type>
+void impl_sum_function_for<Type>::validate_sum() {
```
Comment (espindola) [inline cql3/functions/functions.cc]: We should move all of this code to the .cc file, but having a template declaration in a header and the definition in a .cc file is asking for trouble. Please keep this in the .hh for now.

I will send a RFC patch doing the move and rebase on top of this once it is in.

File: test/boost/cql_query_test.cc [hunk 3]
```diff
@@ -4313,3 +4313,55 @@ SEASTAR_TEST_CASE(test_rf_expand) {
...
+
+        cquery_nofail(e, "insert into cf (pk, val) values ('z', -4294967295);");
+        assert_that(e.execute_cql("select sum(val) from cf;").get0())
+            .is_rows()
+            .with_size(1)
+            .with_row({long_type->decompose(int64_t(-1))});
+    });
+}
+
+SEASTAR_TEST_CASE(test_int_sum_overflow) {
```
Comment (espindola) [inline test/boost/cql_query_test.cc]: This tests already pass is non debug builds, which suggests that std::make_unsigned is the way to go.

File: cql3/functions/functions.cc (hunk 1 above)
Comment (bhalevy) [reply cql3/functions/functions.cc]: Regarding AVG - this agrees with what Cassandra produces (at least for the `int` type, not sure about the `bigint` type, but I'd consider it a bug if the internal accumulator overflows since the average of data of type T must always fit in T)

I can't say that SUM is "defined to be modular", but it behaves this way in Cassandra.
From the user's perspective I think it's bad to lose information about overflow.
Id we suppress any warning then I'd be happy if we add a way in CQL to declare a type type for the SUM that is different than the column type, e.g. `SELECT SUM(val) FROM cf AS bigint;`

Comment (bhalevy) [reply cql3/functions/functions.cc]: @avi / @slivne / @tzach / @nyh  what's you take on the above?

File: cql3/functions/functions.cc (hunk 2 above)
Comment (bhalevy) [reply cql3/functions/functions.cc]: makes sense

File: test/boost/cql_query_test.cc (hunk 3 above)
Comment (bhalevy) [reply test/boost/cql_query_test.cc]: I agree from a programming perspective, but not sure if it's the right thing to do from the usability perspective.

File: cql3/functions/functions.cc (hunk 1 above)
Comment (bhalevy) [reply cql3/functions/functions.cc]: Summoning also @GalitElad

Comment (avikivity) [reply cql3/functions/functions.cc]: Better to throw an exception (and fail the query). There's no point in returning a modulo arithmetic sum.

The user can widen the calculation by using `SELECT SUM(val AS bigint) FROM tab`.

What do other databases do? Perhaps it makes sense for SUM to auto-widen, at least up to bigint.

Comment (bhalevy) [reply cql3/functions/functions.cc]: I'm all for either auto-widen the result (easy), or throw an exception if it doesn't fit (not very hard either)

FWIW, Cassandra decided on silently truncating the result.
See https://issues.apache.org/jira/browse/CASSANDRA-9674

Comment (avikivity) [reply cql3/functions/functions.cc]: Auto-widen is not so easy, because it changes the prepared statement metadata. So we have to introduce a cluster feature, and hash the feature into prepared statement ID so we don't mix an old prepared statement with a new one.

Comment (bhalevy) [reply cql3/functions/functions.cc]: Though I'm not sure how would the clients handle possible overflow in the result.

I suggest that as a first step we throw an error since, as @avikivity mentioned, the user/app can promote the input to safely retrieve a larger result.

Then, as a follow-up, consider auto-promotion of the result type if feasible.

**Comment by slivne** [comment]
I don;t think we can auto widen the result - since the user can be
expecting a specific type returned and as such may have setup his
code accordingly. So we should either throw or truncate.

File: cql3/functions/functions.cc (hunk 1 above)
Comment (avikivity) [reply cql3/functions/functions.cc]: But let's collect information about what other databases do.

**Comment by bhalevy** [comment]
In v3:
- Throw an exception on sum() overflow
- Added more unit tests for sum overflow and overflow-prevention by casting of input to bigint.

File: exceptions/exceptions.hh [hunk 4]
```diff
@@ -203,6 +204,12 @@ struct read_failure_exception : public request_failure_exception {
     { }
 };
 
+class overflow_error_exception: public cassandra_exception {
+public:
+    overflow_error_exception(sstring msg) noexcept
+        : cassandra_exception(exception_code::UNAUTHORIZED, std::move(msg)) {}
```
Comment (avikivity) [inline exceptions/exceptions.hh]: UNAUTHORIZED?

File: exceptions/exceptions.hh [hunk 5]
```diff
@@ -65,6 +65,7 @@ enum class exception_code : int32_t {
     WRITE_FAILURE   = 0x1500,
     READ_TIMEOUT    = 0x1200,
     READ_FAILURE    = 0x1300,
+    OVERFLOW_ERROR  = 0x1600,
```
Comment (avikivity) [inline exceptions/exceptions.hh]: I see this in Cassandra source:

```
src/java/org/apache/cassandra/exceptions/ExceptionCode.java:    CDC_WRITE_FAILURE   (0x1600),
```

Let's allocate our errors in the range 0x1aff-0x1a01 (backwards) to minimize the chance of collision.

Comment (bhalevy) [reply exceptions/exceptions.hh]: sure

File: exceptions/exceptions.hh (hunk 4 above)
Comment (bhalevy) [reply exceptions/exceptions.hh]: sorry, not sure how this came to be this way.

File: exceptions/exceptions.hh (hunk 5 above)
Comment (bhalevy) [reply exceptions/exceptions.hh]: @avi, why `0x1aff-1a01` rather than `0x1fff-0x1f01`?

Comment (avikivity) [reply exceptions/exceptions.hh]: 0x1fff is too special, somebody else might want to use it. 0x1aff is boring and no one would ever pick it on purpose.

**Comment by bhalevy** [comment]
In v4:
- exceptions: define additional cassandra CQL exceptions codes
- Fixed overflow_error_exception code
- Define new OVERFLOW_ERROR exception code as 0x1fff

File: exceptions/exceptions.hh [hunk 6]
```diff
@@ -65,6 +65,8 @@ enum class exception_code : int32_t {
     WRITE_FAILURE   = 0x1500,
     READ_TIMEOUT    = 0x1200,
     READ_FAILURE    = 0x1300,
+    FUNCTION_FAILURE= 0x1400,
```
Comment (espindola) [inline exceptions/exceptions.hh]: Nit for a follow-up: please sort.

File: exceptions/exceptions.hh [hunk 7]
```diff
@@ -68,6 +68,10 @@ enum class exception_code : int32_t {
     FUNCTION_FAILURE= 0x1400,
     CDC_WRITE_FAILURE = 0x1600,
 
+    // Scylla-specific codes
+    // Allocated backwards from 0x1aff-0x1a01 to minimize the chance of collision with Cassandra.
+    OVERFLOW_ERROR  = 0x1aff,
```
Comment (espindola) [inline exceptions/exceptions.hh]: Do we have good enough relations with cassandra developers to propose reserving this code on their end? They don't have to implement the exception now, but if they ever decide to we would at least be compatible.

File: cql3/functions/aggregate_fcts.cc [hunk 8]
```diff
@@ -75,15 +76,66 @@ class count_rows_function final : public native_aggregate_function {
     }
 };
 
+// We need a wider accumulator for sum and average,
```
Comment (espindola) [inline cql3/functions/aggregate_fcts.cc]: The example of 2147483647 + 1 -1 is interesting. With this implementation we promise that we support intermediate values overflowing, as long as the final value is in range.

If we don't want to support that we could use  __builtin_add_overflow. I am OK with both, just checking.

File: test/boost/cql_query_test.cc [hunk 9]
```diff
@@ -4408,3 +4408,28 @@ SEASTAR_TEST_CASE(test_int_sum_with_cast) {
         });
     });
 }
+
+SEASTAR_TEST_CASE(test_bigint_avg) {
```
Comment (espindola) [inline test/boost/cql_query_test.cc]: This test is already passing. Did you intent to use 9223372036854775807 as the constants?

File: cql3/functions/aggregate_fcts.cc (hunk 8 above)
Comment (bhalevy) [reply cql3/functions/aggregate_fcts.cc]: We want to support intermediate overflow to minimize user disruption.

File: exceptions/exceptions.hh (hunk 6 above)
Comment (bhalevy) [reply exceptions/exceptions.hh]: 6ef26a4a4a6de3b0b46bb0b08b083761f90653ee added `WRITE_FAILURE`  disordered so it be close to WRITE_TIMEOUT.
But given the additional values sorting all of them does make more sense.

File: exceptions/exceptions.hh (hunk 7 above)
Comment (bhalevy) [reply exceptions/exceptions.hh]: I definitely don't have these relations.
We can try sending a patch, but I doubt they'll accept it with no implemented use of this value.

File: test/boost/cql_query_test.cc (hunk 9 above)
Comment (bhalevy) [reply test/boost/cql_query_test.cc]: yes, probably copy/paste error. good catch.
I'll fix that and resend a new version (including sorting of the error codes).

**Comment by bhalevy** [comment]
In v5:
- Added "exceptions: sort exception_code definitions"
- Fixed test_bigint_avg to induce overflow (pre-fix) using max bigint (9223372036854775807)

File: cql3/functions/aggregate_fcts.cc (hunk 8 above)
Comment (avikivity) [reply cql3/functions/aggregate_fcts.cc]: Certainly it's important for average. avg(tinyint) always fits in a tinyint, but the intermediate accumulator can easily overflow.

For sum() it's not important, but we have it for free so why not.

File: exceptions/exceptions.hh (hunk 7 above)
Comment (avikivity) [reply exceptions/exceptions.hh]: They won't, they made it clear in another occasion.
